consensus/polybft/extra.go

```diff
@@ -0,0 +1,390 @@
+		return fmt.Errorf("incorrect elements count to decode validator set delta, expected 2 but found %d", num)
+	}
+
+	// Validators (added)
+	{
+		validatorsRaw, err := elems[0].GetElems()
+		if err != nil {
+			return fmt.Errorf("array expected for added validators")
+		}
+
+		if len(validatorsRaw) != 0 {
+			d.Added = make(AccountSet, len(validatorsRaw))
+
+			for i, validatorRaw := range validatorsRaw {
+				acc := &ValidatorAccount{}
+				if err = acc.UnmarshalRLPWith(validatorRaw); err != nil {
+					return err
+				}
+
+				d.Added[i] = acc
+			}
+		}
+	}
+
+	// Bitmap (removed)
+	{
+		dst, err := elems[1].GetBytes(nil)
+		if err != nil {
+			return err
+		}
+
+		d.Removed = bitmap.Bitmap(dst)
+	}
+
+	return nil
+}
+
+// IsEmpty returns indication whether delta is empty (namely added and removed slices are empty)
+func (d *ValidatorSetDelta) IsEmpty() bool {
+	return len(d.Added) == 0 && d.Removed.Len() == 0
+}
+
+// Copy creates deep copy of ValidatorSetDelta
+func (d *ValidatorSetDelta) Copy() *ValidatorSetDelta {
+	added := d.Added.Copy()
+	removed := make([]byte, len(d.Removed))
+	copy(removed, d.Removed)
+
+	return &ValidatorSetDelta{Added: added, Removed: removed}
+}
+
+// fmt.Stringer interface implementation
+func (d *ValidatorSetDelta) String() string {
+	return fmt.Sprintf("Added %v Removed %v", d.Added, d.Removed)
+}
+
+// Signature represents aggregated signatures of signers accompanied with a bitmap
+// (in order to be able to determine identities of each signer)
+type Signature struct {
+	AggregatedSignature []byte
+	Bitmap              []byte
+}
+
+// MarshalRLPWith marshals Signature object into RLP format
+func (s *Signature) MarshalRLPWith(ar *fastrlp.Arena) *fastrlp.Value {
+	committed := ar.NewArray()
+	if s.AggregatedSignature == nil {
+		committed.Set(ar.NewNull())
+	} else {
+		committed.Set(ar.NewBytes(s.AggregatedSignature))
+	}
+
+	if s.Bitmap == nil {
+		committed.Set(ar.NewNull())
+	} else {
+		committed.Set(ar.NewBytes(s.Bitmap))
+	}
+
+	return committed
+}
+
+// UnmarshalRLPWith unmarshals Signature object from the RLP format
+func (s *Signature) UnmarshalRLPWith(v *fastrlp.Value) error {
+	vals, err := v.GetElems()
+	if err != nil {
+		return fmt.Errorf("array type expected for signature struct")
+	}
+
+	// there should be exactly two elements (aggregated signature and bitmap)
+	if len(vals) != 2 {
+		return fmt.Errorf("invalid rlp values")
+	}
+
+	s.AggregatedSignature, err = vals[0].GetBytes(nil)
+	if err != nil {
+		return err
+	}
+
+	s.Bitmap, err = vals[1].GetBytes(nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// VerifyCommittedFields is checking for consensus proof in the header
+func (s *Signature) VerifyCommittedFields(validatorSet AccountSet, hash types.Hash) error {
+	filtered, err := validatorSet.GetFilteredValidators(s.Bitmap)
+	if err != nil {
+		return err
+	}
+
+	if len(filtered) < getQuorumSize(validatorSet.Len()) {
+		return fmt.Errorf("quorum not reached")
```

Comment:
This can be an errors.New, instead of a formatted error (nothing is being formatted)